frontend/catalyst/utils/wrapper.cpp

```diff
@@ -157,76 +158,81 @@ py::list move_returns(void *memref_array_ptr, py::object result_desc, py::object
             throw std::runtime_error("PyArray_NewFromDescr failed.");
         }
 
-        PyObject *capsule =
-            PyCapsule_New(memref->allocated, NULL, (PyCapsule_Destructor)&free_wrap);
+        PyObject *capsule = PyCapsule_New(memref->allocated, NULL,
+                                          reinterpret_cast<PyCapsule_Destructor>(&free_wrap));
         if (!capsule) {
             throw std::runtime_error("PyCapsule_New failed.");
         }
 
-        int retval = PyArray_SetBaseObject((PyArrayObject *)new_array, capsule);
+        int retval = PyArray_SetBaseObject(reinterpret_cast<PyArrayObject *>(new_array), capsule);
         bool success = 0 == retval;
         if (!success) {
             throw std::runtime_error("PyArray_SetBaseObject failed.");
         }
 
-        returns.append(new_array);
+        returns.append(nb::borrow(new_array));
```

Comment:
Just curious if you could explain the use of borrow :)

Reply:
I'm glad you pointed this out because I think that we should actually use [`nb::steal`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4I0EN8nanobind5stealE1T6handle) here, not [`nb::borrow`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4I0EN8nanobind6borrowE1T6handle).

The short answer is that without one of `nb::borrow` or `nb::steal`, you get a `std::bad_cast` runtime error!

Here, `new_array` is a raw [`PyObject*`](https://docs.python.org/3/c-api/structures.html#c.PyObject) pointer and `returns` is a [`nb::list`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4N8nanobind4listE) (a wrapper class representing a Python `list`). In order to call [`append()`](https://nanobind.readthedocs.io/en/latest/api_core.html#_CPPv4I0EN8nanobind4list6appendEvRR1T) on `returns`, its argument must be a wrapped Python object. To wrap the `PyObject*` pointer, we have the choice of `nb::steal`, which *steals* a reference to the object (and does **not** increment its reference count), or `nb::borrow` which *borrows* a reference to the object (and **does** increments its reference count). Stealing is generally for C API calls that return a new object that we now own and must manage, and borrowing is for C API calls that return a (borrowed) reference to an object that another part of the program owns and manages.

The documentation for [`PyArray_NewFromDescr`](https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_NewFromDescr) says that it steals a reference to `descr` (which makes sense since it's returning a *new* array), so we should use `nb::steal` here.

Reply:
Nice, thanks for explaining it :)

@erick-xanadu Does that makes sense to you since you wrote the original?

Reply:
I think borrow makes sense given that we manually decrement the reference in line 191. Have you tried stealing and everything worked correctly?

Reply:
If it works with borrow, then that should be enough proof too. You can manually [look at the reference count](https://docs.python.org/3/c-api/refcounting.html#c.Py_REFCNT). Essentially at the end of this method, the reference count should be two. One for the list and one for the dictionary.

Reply:
Ah, so we could either borrow and decrement, or steal?

Reply:
> Ah, so we could either borrow and decrement, or steal?

I think we need to `(borrow and decrement) xor steal`

Reply:
Thanks @dime10 and @erick-xanadu! I hadn't considered that we insert `new_array` in two places. I printed out the reference counts of `new_array` just to be sure; I get:

1. Increment ref count to 1 after array creation
2. Increment ref count to 2 after `returns.append(nb::borrow(new_array))`
3. Increment ref count to 3 after `numpy_arrays[pyLong] = nb::borrow(new_array)`
4. Decrement ref count to 2 with `Py_DECREF(new_array)`

If we were using the array once it might make more sense to use `nb::steal` so we don't have to manually decrement the ref count. We could steal once and borrow once, but that might looks strange and confuse anyone looking at this in the future. I propose we keep it as is with two borrows and a final `Py_DECREF`. I'll add a comment saying that the ref count of `new_array` at the end should be 2.